be/src/vec/functions/array/function_array_compact.h

```diff
@@ -0,0 +1,251 @@
+                        size_t result, size_t input_rows_count) override {
+        ColumnPtr src_column =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
+        const auto& src_column_array = check_and_get_column<ColumnArray>(*src_column);
+        if (!src_column_array) {
+            return Status::RuntimeError(
+                    fmt::format("unsupported types for function {}({})", get_name(),
+                                block.get_by_position(arguments[0]).type->get_name()));
+        }
+        const auto& src_offsets = src_column_array->get_offsets();
+        const auto* src_nested_column = &src_column_array->get_data();
+        DCHECK(src_nested_column != nullptr);
+
+        DataTypePtr src_column_type = block.get_by_position(arguments[0]).type;
+        auto nested_type = assert_cast<const DataTypeArray&>(*src_column_type).get_nested_type();
+        auto dest_column_ptr = ColumnArray::create(nested_type->create_column(),
+                                                   ColumnArray::ColumnOffsets::create());
+        IColumn* dest_nested_column = &dest_column_ptr->get_data();
+        auto& dest_offsets = dest_column_ptr->get_offsets();
+        DCHECK(dest_nested_column != nullptr);
+
+        const NullMapType* src_null_map = nullptr;
+        if (src_nested_column->is_nullable()) {
+            const ColumnNullable* src_nested_nullable_col =
+                    check_and_get_column<ColumnNullable>(*src_nested_column);
+            src_nested_column = src_nested_nullable_col->get_nested_column_ptr();
+            src_null_map = &src_nested_nullable_col->get_null_map_column().get_data();
+        }
+
+        NullMapType* dest_null_map = nullptr;
+        if (dest_nested_column->is_nullable()) {
+            ColumnNullable* dest_nested_nullable_col =
+                    reinterpret_cast<ColumnNullable*>(dest_nested_column);
+            dest_nested_column = dest_nested_nullable_col->get_nested_column_ptr();
+            dest_null_map = &dest_nested_nullable_col->get_null_map_column().get_data();
+        }
+
+        auto res_val = _execute_by_type(*src_nested_column, src_offsets, *dest_nested_column,
+                                        dest_offsets, src_null_map, dest_null_map, nested_type);
+        if (!res_val) {
+            return Status::RuntimeError(
+                    fmt::format("execute failed or unsupported types for function {}({})",
+                                get_name(), block.get_by_position(arguments[0]).type->get_name()));
+        }
+
+        block.replace_by_position(result, std::move(dest_column_ptr));
+        return Status::OK();
+    }
+
+private:
+    template <typename ColumnType>
+    bool _execute_number(const IColumn& src_column, const ColumnArray::Offsets64& src_offsets,
+                         IColumn& dest_column, ColumnArray::Offsets64& dest_offsets,
```

Comment:
why not we just use `ColumnArray's nested column` as input arguments, and then we could ignore nullable situation, the code is more like 
```
++src_pos;
++res_pos;
for (; src_pos < src_offset; ++src_pos) {
    if (nested_src->compareAt(src_pos - 1, src_pos, *nested_src, 1)) {
        nested_dst_column->insert_from(src_values, src_pos);
        ++res_pos;
    }
}
```

if `nested_dst_column` and `nested_src` is nullable, then insert_from will handle whether the value is null or not automatically， so you won't need to take care of the null values

Reply:
the code  could be more clean and clear

Reply:
Yeah, you are right and I will refactor this part of the code soon.